File: Problem/BOJ_2583_영역 구하기/main.js
const fs = require('fs')
const input = fs.readFileSync("./input.txt").toString().split('\n')

const [rows, cols, numOfSquares] = input[0].split(" ").map(elem => parseInt(elem))

const movingDir = [[-1, 0], [1, 0], [0, -1], [0, 1]]

const graphSquare = Array.from(Array(rows), () => Array(cols).fill(0))
input.slice(1).forEach(positions => {
    const [x1, y1, x2, y2] = positions.split(" ").map(elem => parseInt(elem))
    for (let i = rows - y2; i < rows - y1; i++) {
        for (let j = x1; j < x2; j++) {
            graphSquare[i][j] = 1
        }
    }
})


function dfs(graphPaper, rows, cols, curRow, curCol, areas) {
    graphPaper[curRow][curCol] = -1

    for (let i = 0; i < movingDir.length; i++) {
        const [nextRow, nextCol] = [curRow + movingDir[i][0], curCol + movingDir[i][1]]
        if (0 <= nextRow && nextRow < rows && 0 <= nextCol && nextCol < cols) {
            if (graphPaper[nextRow][nextCol] === 0) {
                areas = dfs(graphPaper, rows, cols, nextRow, nextCol, areas) + 1
            }
        }
    }
    return areas
}

let numOfAreas = 0
const areaList = []
for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
        if (graphSquare[row][col] === 0) {
            numOfAreas += 1
            areaList.push(dfs(graphSquare, rows, cols, row, col, 1))
        }
    }
}

console.log(numOfAreas)
console.log(areaList.sort((a, b) => a - b).join(" "))

File: Problem/BOJ_2583_영역 구하기/main.py
from sys import stdin, setrecursionlimit
setrecursionlimit(10 ** 5)

moving_dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]


def dfs(cur_row, cur_col, area):
    visited[cur_row][cur_col] = True

    for moving_row, moving_col in moving_dir:
        next_row = cur_row + moving_row
        next_col = cur_col + moving_col
        if 0 <= next_row < rows and 0 <= next_col < cols:
            if paper[next_row][next_col] == 0 and not visited[next_row][next_col]:
                area = dfs(next_row, next_col, area) + 1

    return area


def main():
    stdin = open("./input.txt", "r")
    global rows, cols, num_of_squares, paper, visited
    rows, cols, num_of_squares = map(int, stdin.readline().split())

    paper = [[0] * cols for _ in range(rows)]

    for _ in range(num_of_squares):
        x1, y1, x2, y2 = map(int, stdin.readline().split())

        for row in range(rows - y2, rows - y1):
            for col in range(x1, x2):
                paper[row][col] = 1

    # print(paper)
    visited = [[False] * cols for _ in range(rows)]

    num_of_area = 0
    areas = []
    for row in range(rows):
        for col in range(cols):
            if paper[row][col] == 0 and not visited[row][col]:
                num_of_area += 1
                areas.append(dfs(row, col, 1))

    print(num_of_area)

    areas.sort()
    print(' '.join(map(str, areas)))


if __name__ == '__main__':
    main()